### PR TITLE
♻️ HOCS-4383: Create a CaseDataTypeService

### PR DESCRIPTION
**Pull request by nickmcmahon01**
As part of decoupling our microservices we need to start to treat data that is currently mastered in Info service as part of the casework domain.
This PR is a refactor to create a CaseDataTypeService which initially retrieves data from Info service and later will be changed to use a repository class without having to change the rest of the code. This code has moved from CaseDataService and StageService.

### CODE REVIEW

File: src/main/java/uk/gov/digital/ho/hocs/casework/client/infoclient/InfoClient.java [hunk 1]
```diff
@@ -30,24 +30,9 @@ public InfoClient(RestHelper restHelper,
         this.serviceBaseURL = infoService;
     }
```
Comment (nickmcmahon01) [inline src/main/java/uk/gov/digital/ho/hocs/casework/client/infoclient/InfoClient.java:32]: We can reduce this to one call that fetches all caseDataTypes - we can present it differently in the service layer. it also allows us to deprecate or remove these calls from info-service.

File: src/main/java/uk/gov/digital/ho/hocs/casework/migration/MigrationCaseDataService.java [hunk 2]
```diff
@@ -56,7 +56,7 @@ CaseData createCase(MigrationCreateCaseRequest request) {
         String totalsListName = request.getTotalsListName();
 
         log.debug("Creating Case of type: {}", caseType);
-        CaseDataType caseDataType = infoClient.getCaseType(caseType);
+        CaseDataType caseDataType = null;//infoClient.getCaseType(caseType);
```
Comment (nickmcmahon01) [inline src/main/java/uk/gov/digital/ho/hocs/casework/migration/MigrationCaseDataService.java:59]: 👎

File: src/test/java/uk/gov/digital/ho/hocs/casework/api/CaseDataCreateCaseIntegrationTest.java [hunk 3]
```diff
@@ -70,7 +70,7 @@ public void shouldCreateACaseWithPermissionLevelOwner() throws JsonProcessingExc
         long numberOfCasesBefore = caseDataRepository.count();
 
         setupMockTeams("TEST", 5);
```
Comment (nickmcmahon01) [inline src/test/java/uk/gov/digital/ho/hocs/casework/api/CaseDataCreateCaseIntegrationTest.java:72]: Boyscout changes, we were passing in these params and not using it them, it looks like they've moved to setupMockTeams above and never been removed from this method params

File: src/test/java/uk/gov/digital/ho/hocs/casework/api/CaseDataCreateCaseIntegrationTest.java [hunk 4]
```diff
@@ -296,9 +296,9 @@ private void setupMockTeams(String caseType, int permission) throws JsonProcessi
         teamDtos.add(teamDto);
 
         mockInfoService
-                .expect(requestTo("http://localhost:8085/caseType/type/TEST"))
+                .expect(requestTo("http://localhost:8085/caseType"))
```
Comment (nickmcmahon01) [inline src/test/java/uk/gov/digital/ho/hocs/casework/api/CaseDataCreateCaseIntegrationTest.java:299]: Most of the real test changes are repointing wiremock to use the one remaining endpoint.

File: src/test/java/uk/gov/digital/ho/hocs/casework/api/CaseDataServiceTest.java [hunk 5]
```diff
@@ -152,7 +138,7 @@ public void shouldCreateCase() throws ApplicationExceptions.EntityCreationExcept
 
         when(caseDataRepository.getNextSeriesId()).thenReturn(caseID);
         when(infoClient.getCaseDeadline(caseType.getDisplayCode(), deadlineDate, 0)).thenReturn(caseDeadline);
-        when(infoClient.getCaseType(caseType.getDisplayCode())).thenReturn(caseType);
```
Comment (nickmcmahon01) [inline src/test/java/uk/gov/digital/ho/hocs/casework/api/CaseDataServiceTest.java:155]: Where we were calling infoClient directly we have repointed the code to call the new CaseDataTypeService.

File: src/test/java/uk/gov/digital/ho/hocs/casework/api/CaseDataServiceTest.java [hunk 6]
```diff
@@ -899,42 +885,6 @@ public void shouldNotDeleteCaseMissingCaseUUID() {
         verifyNoMoreInteractions(caseDataRepository);
     }
 
-    @Test
```
Comment (nickmcmahon01) [inline src/test/java/uk/gov/digital/ho/hocs/casework/api/CaseDataServiceTest.java:902]: These have been removed from CaseDataServiceTest and moved to CaseDataTypeServiceTest

File: src/test/java/uk/gov/digital/ho/hocs/casework/security/AuthorisationAspectTest.java [hunk 7]
```diff
@@ -58,167 +58,149 @@ public void setup() {
...
         request.setMethod("GET");
         RequestContextHolder.setRequestAttributes(new ServletRequestAttributes(request));
-        aspect = new AuthorisationAspect(caseService, userService);
+        aspect = new AuthorisationAspect(caseService, caseDataTypeService, userService);
     }
 
     @Test
     public void shouldCaseServiceLookupWhenExistingCase() throws Throwable {
 
-        String type = "MIN";
```
Comment (nickmcmahon01) [inline src/test/java/uk/gov/digital/ho/hocs/casework/security/AuthorisationAspectTest.java:67]: The gitblame is me, but I don't like using real data in tests where possible, these tests are not DCU specific behaviour

File: src/main/java/uk/gov/digital/ho/hocs/casework/api/CaseDataTypeService.java [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+    private final InfoClient infoClient;
+
+    @Autowired
+    public CaseDataTypeService(InfoClient infoClient) {
+        this.infoClient = infoClient;
+    }
+
+    public Stream<CaseDataType> getAllCaseDataTypes(Predicate<CaseDataType> predicate) {
+        return infoClient.getAllCaseDataTypes().stream().filter(predicate);
```
Comment (LiamMacP) [inline src/main/java/uk/gov/digital/ho/hocs/casework/api/CaseDataTypeService.java:29]: Part of me sort of feels like the Predicate stuff shouldn't be part of this method, `getAllCaseDataTypes` aludes to returning them all...

Comment (LiamMacP) [reply src/main/java/uk/gov/digital/ho/hocs/casework/api/CaseDataTypeService.java:29]: I do however really like us using a properly cached method and pulling the value from that...

File: src/main/java/uk/gov/digital/ho/hocs/casework/api/CaseDataTypeService.java [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+    }
+
+    public CaseDataType getCaseDataType(String caseType) {
+        var type = getNonNullValue(caseType);
+        return getAllCaseDataTypes(it -> it.getDisplayCode().equals(type)).findFirst().orElseThrow(() -> getNotFoundException(type));
+    }
+
+    public CaseDataType getCaseDataType(UUID caseUUID){
+        var shortCode = getNonNullValue(caseUUID).toString().substring(34);
```
Comment (LiamMacP) [inline src/main/java/uk/gov/digital/ho/hocs/casework/api/CaseDataTypeService.java:38]: Hmm, can we not move away from the weird manipulation of UUIDs? I know why they are there, it just feels wrong though...

File: src/main/java/uk/gov/digital/ho/hocs/casework/api/CaseDataTypeService.java [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+        return getAllCaseDataTypes(it -> it.getShortCode().equals(shortCode)).findFirst().orElseThrow(() -> getNotFoundException(shortCode));
+    }
+
+    private static <T> T getNonNullValue(T value) {
+        return Optional.ofNullable(value).orElseThrow(() -> getNotFoundException("null"));
```
Comment (LiamMacP) [inline src/main/java/uk/gov/digital/ho/hocs/casework/api/CaseDataTypeService.java:43]: Wouldn't this always just lead to `CaseDataType: null, not found!`, not sure thats much use...

File: src/main/java/uk/gov/digital/ho/hocs/casework/api/TopicResource.java [hunk 11]
```diff
@@ -56,15 +56,9 @@ ResponseEntity deleteTopic(@PathVariable UUID caseUUID, @PathVariable UUID stage
         return ResponseEntity.ok().build();
     }
 
-    @PostMapping(value = "/topic/{topicUUID}/clearCachedStandardLine")
-    ResponseEntity clearCachedStandardLineForTopic(@PathVariable UUID topicUUID) {
```
Comment (LiamMacP) [inline src/main/java/uk/gov/digital/ho/hocs/casework/api/TopicResource.java:60]: Glad to see this go if unused...

File: src/test/java/uk/gov/digital/ho/hocs/casework/api/CaseDataCreateCaseIntegrationTest.java [hunk 12]
```diff
@@ -48,7 +48,7 @@
 @ActiveProfiles({ "local", "integration" })
 public class CaseDataCreateCaseIntegrationTest {
     private MockRestServiceServer mockInfoService;
-    private TestRestTemplate testRestTemplate = new TestRestTemplate();
+    private final TestRestTemplate testRestTemplate = new TestRestTemplate();
```
Comment (LiamMacP) [inline src/test/java/uk/gov/digital/ho/hocs/casework/api/CaseDataCreateCaseIntegrationTest.java:51]: 👮🏽‍♂️

File: src/test/java/uk/gov/digital/ho/hocs/casework/api/StageServiceTest.java [hunk 13]
```diff
@@ -64,7 +66,7 @@
     private final String allocationType = "anyAllocate";
     private final UUID transitionNoteUUID = UUID.randomUUID();
     private final CaseDataType caseDataType = new CaseDataType("MIN", "1a", "MIN", null);
-    private final List<CaseDataType> caseDataTypes = List.of(
+    private final Stream<CaseDataType> caseDataTypes = Stream.of(
             CaseDataTypeFactory.from("NXT", "a5", "MIN"), // NXT can be reached through MIN
```
Comment (LiamMacP) [inline src/test/java/uk/gov/digital/ho/hocs/casework/api/StageServiceTest.java:70]: Not really related but can we change `MIN` to something not tenant specifc?

File: src/test/java/uk/gov/digital/ho/hocs/casework/security/SecurityIntegrationTest.java [hunk 14]
```diff
@@ -53,6 +54,9 @@
     @Qualifier("CaseDataService")
     CaseDataService caseDataService;
 
+    @MockBean()
```
Comment (LiamMacP) [inline src/test/java/uk/gov/digital/ho/hocs/casework/security/SecurityIntegrationTest.java:57]:
```suggestion
    @MockBean
```

**Review by LiamMacP — DISMISSED** [review]
Nothing that cannot be re-worked going forward, good improvement overall...

File: src/main/java/uk/gov/digital/ho/hocs/casework/api/CaseDataTypeService.java (hunk 9 above)
Comment (nickmcmahon01) [reply src/main/java/uk/gov/digital/ho/hocs/casework/api/CaseDataTypeService.java:38]: I agree, but it needs fixing upstream of here, for the security code we can stop looking up the caseType completely and fetch the actual case from the database, which is quick enough, but then semantically there is a difference 'if it exists this UUID would be a MIN' to 'this case isn't in the database' so we would need to ensure the behaviour stays the same.

**Pull request closed without merging**